clash-lib/src/Clash/Netlist/BlackBox.hs

```diff
@@ -455,28 +455,40 @@ mkFunInput
 mkFunInput resId e =
  let (appE,args,ticks) = collectArgsTicks e
  in  withTicks ticks $ \tickDecls -> do
+  tcm <- Lens.use tcCache
   -- TODO: Rewrite this function to use blackbox functions. Right now it
   -- TODO: generates strings that are later parsed/interpreted again. Silly!
   (bbCtx,dcls) <- mkBlackBoxContext "__INTERNAL__" resId (lefts args)
   templ <- case appE of
             Prim nm _ -> do
               bb  <- extractPrimWarnOrFail nm
-              let
-                templ =
-                  case bb of
-                    P.BlackBox {..} ->
-                      Left (kind,outputReg,libraries,imports,includes,nm,template)
-                    P.Primitive pn _ pt ->
-                      error $ $(curLoc) ++ "Unexpected blackbox type: "
-                                        ++ "Primitive " ++ show pn
-                                        ++ " " ++ show pt
-                    P.BlackBoxHaskell pnm _ fn _ ->
-                      error $ $(curLoc) ++ "Unexpected blackbox type: "
-                                        ++ "BlackBoxHaskell" ++ show pnm
-                                        ++ " " ++ show fn ++ " <func>"
-              return templ
+              case bb of
+                P.BlackBox {..} ->
+                  pure (Left (kind,outputReg,libraries,imports,includes,nm,template))
+                P.Primitive pn _ pt ->
+                  error $ $(curLoc) ++ "Unexpected blackbox type: "
+                                    ++ "Primitive " ++ show pn
+                                    ++ " " ++ show pt
+                P.BlackBoxHaskell pName _workInfo fName (_, func) -> do
+                  -- Determine result type of this blackbox. If it's not a
+                  -- function, simply use its term type.
+                  let
+                    resTy0 = termType tcm e
+                    resTy1 =
+                      case splitFunTy tcm resTy0 of
+                        Just (_, t) -> t
+                        Nothing -> resTy0
+
+                  bbhRes <- func True (Right resId) pName args resTy1
```

Comment:
Point of attention: I'm not sure whether passing in `(Right resId)` is the correct thing to do here, but I'm not sure what else to do..?

Reply:
No... this isn't right, that `resId` is the lhs in:

```haskell
lhs = map fun xs
```

While the BlackBoxHaskell is expecting an identifier to bind the result of `fun`.

Reply:
Right, that's what I suspected. Should we just remove the argument and let users use `~RESULT`?

Reply:
Actually, I'm pretty sure that's the only valid option here, as there is no singular result binder id. It depends on the instantiation of the blackbox later on.

Reply:
Yeah... knowing the type of the result should be enough.